### ISSUE TITLE
[BR]: Error message "string index out of range"

### ISSUE THREAD

**Issue by ronyshtamler** [fail2ban/fail2ban#2967]
### Environment:

- Fail2Ban version  : 
- OS, including release name/version : 
- [X] Fail2Ban installed via OS/distribution mechanisms
- [X] You have not applied any additional foreign patches to the codebase
- [ ] Some customizations were done to the configuration (provide details below is so)

### The issue:
Version 11.2 running under docker

Many error messages and eventually Fail2ban crashes:
```
2021-03-11 14:21:05,710 fail2ban.filter         [1]: ERROR   Failed to process line: '2f06fefa-5f41-4bcb-a8c9-7de085a9a74f \t[C]: in function rename', caught exception: IndexError('string index out of range')

2021-03-11 14:40:47,208 fail2ban.filter         [1]: ERROR   Failed to process line: '09d44a7a-4fff-463b-a735-375231809cc5 stack traceback:', caught exception: IndexError('string index out of range')
```

**Comment by sebres** [comment]
> Many error messages ...

People!... our issue template tells you which information do we need to be able to process the issue.
Can you please provide which jail it does affect, how the one or other log line does look in reality (in original log), and if possibly the excerpt of `fail2ban-client --dp` (or filter file definition) with all regex's for this jail?

> and eventually Fail2ban crashes

What does mean eventually? A crash is unambiguous, either it does or it doesn't.
For instance the process ID should change in log, and you would see `Starting Fail2ban v.11.2` in the fail2ban.log file (if it is automatically restarting).

**Comment by sebres** [comment]
Possible duplicate of #2933 etc.
Are in monitored log-file some timestamps from year 2020 or before?
If so it is already fixed, see https://github.com/fail2ban/fail2ban/issues/2904#issuecomment-754042663

**Comment by ronyshtamler** [comment]
> > and eventually Fail2ban crashes
> 
> What does mean eventually? A crash is unambiguous, either it does or it doesn't.
> For instance the process ID should change in log, and you would see `Starting Fail2ban v.11.2` in the fail2ban.log file (if it is automatically restarting).

The message is:
2021-03-11 06:30:05,529 fail2ban.filter         [1]: ERROR   Too many errors at once (100), going idle

**Comment by ronyshtamler** [comment]
> > Many error messages ...
> 
> People!... our issue template tells you which information do we need to be able to process the issue.
> Can you please provide which jail it does affect, how the one or other log line does look in reality (in original log), and if possibly the excerpt of `fail2ban-client --dp` (or filter file definition) with all regex's for this jail?
> 
Please excuse my lack of following the instructions opening this issue. I will add more information once I figure out which jail is responsible.

**Comment by ronyshtamler** [comment]
I found the problem.
it was an error in the log file - on occasion it sput up an unfinished buffer instead of the timestamp.

**Comment by sebres** [comment]
Good, glad it works now. By the way...
> Too many errors at once (100), going idle

means the jail (affected by this errors) only becomes idle, neither it is a crash, nor fail2ban stops to work (in other jails).

**Comment by sebres** [comment]
> it sput up an unfinished buffer instead of the timestamp.

However, can you provide an example, just to ensure fail2ban can get some preventive mechanism in order to avoid such errors in the feature?

**Comment by brianjmurrell** [comment]
I am getting this also.  Log lines such as:

```
Jun 26 03:13:25 server logrotate[1287146]: ALERT exited abnormally with [1]
Jun 26 03:13:25 server named-pkcs11[3645]: zone 101.168.192.in-addr.arpa/IN: loaded serial 1656227570
Jun 26 03:13:30 server named-pkcs11[3645]: zone 0.0.0.0.f.d.8.4.1.b.e.a.1.3.d.f.ip6.arpa/IN: sending notifies (serial 1656227568)
```

Seem to be among those causing it.

Jail definition:

```
['add', 'gw-portscan', 'auto']
['set', 'gw-portscan', 'usedns', 'warn']
['multi-set', 'gw-portscan', 'addfailregex', ['net2(.+):DROP:.* SRC=(?:::f{4,6}:)?(<HOST>) .*(PROTO=TCP .*SYN|PROTO=UDP)', '(net2net|sfilter):DROP:.* SRC=<HOST>']]
['set', 'gw-portscan', 'maxretry', 3]
['set', 'gw-portscan', 'maxmatches', 3]
['set', 'gw-portscan', 'findtime', '86400']
['set', 'gw-portscan', 'bantime', '10m']
['set', 'gw-portscan', 'ignorecommand', '']
['set', 'gw-portscan', 'addignoreip', '127.0.0.1/8', '::1', '99.234.253.96/32', '205.189.48.131/32', '184.75.215.106/32', '138.197.170.241/32', '50.116.16.106/32', '52.48.88.88/32']
['set', 'gw-portscan', 'logencoding', 'auto']
['set', 'gw-portscan', 'addlogpath', '/var/log/messages', 'head']
['set', 'gw-portscan', 'addaction', 'ipset-restore']
['multi-set', 'gw-portscan', 'action', 'ipset-restore', [['actionstart', ''], ['actionstop', ''], ['actioncheck', ''], ['actionban', 'echo "ipset -A <setname> <ip>" >> /var/lib/fail2ban/ipset-restore'], ['actionunban', "sed -i -e>
['set', 'gw-portscan', 'addaction', 'log-ban-PORTSCAN']
['multi-set', 'gw-portscan', 'action', 'log-ban-PORTSCAN', [['actionstart', ''], ['actionstop', ''], ['actioncheck', ''], ['actionban', 'echo "$(date \'+%Y-%m-%d %T\') <ip>" >> /var/log/fail2ban/bans.PORTSCAN'], ['actionunban', '>
['set', 'gw-portscan', 'addaction', 'gw-shorewall-PORTSCAN']
['multi-set', 'gw-portscan', 'action', 'gw-shorewall-PORTSCAN', [['actionstart', ''], ['actionstop', ''], ['actionflush', 'true'], ['actioncheck', ''], ['actionban', 'echo <ip> | /root/bin/action-ban drop fail2ban <name>'], ['act>
['set', 'gw-portscan', 'addaction', 'sendmail-lines-whois-PORTSCAN']
[ 'multi-set',
  'gw-portscan',
  'action',
  'sendmail-lines-whois-PORTSCAN',
  [ ['actionstart', 'printf %b "Subject: [Fail2Ban] PORTSCAN: started on <fq-hostname>\nDate: `LC_ALL=C date +"%a, %d %h %Y %T %z"`\nFrom: Fail2Ban <root@example.com>\nTo: root\\n\nHi,\\n\nThe jail PORTSCAN has been started s>
    ['actionstop', 'printf %b "Subject: [Fail2Ban] PORTSCAN: stopped on <fq-hostname>\nDate: `LC_ALL=C date +"%a, %d %h %Y %T %z"`\nFrom: Fail2Ban <root@example.com>\nTo: root\\n\nHi,\\n\nThe jail PORTSCAN has been stopped.\\>
    ['actioncheck', ''],
    ['actionban', 'printf %b "Subject: [Fail2Ban] PORTSCAN: banned <ip> from <fq-hostname>\nDate: `LC_ALL=C date +"%a, %d %h %Y %T %z"`\nFrom: Fail2Ban <root@example.com>\nTo: root\\n\nHi,\\n\nThe IP <ip> has just been banned>
    ['actionunban', ''],
    ['norestored', True],
    ['name', 'PORTSCAN'],
    ['dest', 'root'],
    ['sender', 'root@example.com'],
    ['logpath', '/var/log/messages'],
    ['actname', 'sendmail-lines-whois-PORTSCAN'],
    ['mailcmd', '/usr/sbin/sendmail -f "<sender>" "<dest>"'],
    ['sendername', 'Fail2Ban']]]
```

**Comment by brianjmurrell** [comment]
Any idea what is going on here with the log messages and config I posted above?  This is preventing the jail that gets the most hits on my system from starting.

**Comment by sebres** [comment]
> Any idea what is going on here with the log messages and config I posted above? 

I don't think it is a config or even the log-format.
Such issues occurred only if the `datepattern` was unspecified (so it trying to apply some previously found default datepattern match), the date is not there (or the message is not written in one piece) or the message is too short (without the datetime).
Another variant would be #2904 (but if I see your data I don't think it is affecting you).

Anyway I cannot really see where 0.11.2 can really be affected by this log-excerpt and the jail.
Please provide exact error message from fail2ban.log.

**Comment by brianjmurrell** [comment]
Hrm.  I see I was supposed to provide an exact error message.  Here is one of many:

```
2022-12-10 10:41:25,213 fail2ban.filter         [2162]: ERROR   Failed to process line: 'Dec  4 03:08:46 server named-pkcs11[3556]: checkhints: unable to get root NS rrset from cache: not found', caught exception: IndexError('string index out of range',)
```

**Comment by sebres** [comment]
It is strange. Neither I think 0.11.2 is affected by the issue at all, nor the messages from this log excerpt could cause the issue, since an anchored datepattern always win (and doesn't seek for other pattern, which may be `Epoch`, what has indeed dynamic position in log-lines, but as already said not interesting since an anchored pattern `{^LN-BEG}(?:%a )?%b %d %k:%M:%S(?:\.%f)?(?: %ExY)?`).

So either it is something different or the constellation is still unknown.

Anyway it is hard to estimate what exactly happens without debugging.
If it is pretty reproducible possibly an attempt with a higher log-level would help (`DEBUG` or `HEAVYDEBUG`, just they are too wordy in fail2ban log). Alternatively with a small patch for https://github.com/fail2ban/fail2ban/blob/0.11.2/fail2ban/server/filter.py#L707-L708:
```diff
  			logSys.error("Failed to process line: %r, caught exception: %r", line, e,
- 				exc_info=logSys.getEffectiveLevel()<=logging.DEBUG)
+ 				exc_info=True)

```
(then it would always log the callstack for that error, no matter which log-level is set)
